crates/primitives/src/checkpoints.rs

```diff
@@ -121,6 +123,88 @@ pub struct StorageHashingCheckpoint {
     pub to: u64,
 }
 
+/// Saves the progress of a stage.
+#[main_codec]
+#[derive(Debug, Default, PartialEq, Eq, Clone, Copy)]
+pub struct StageCheckpoint {
+    /// Stage progress measured in block number it reached.
+    pub block_number: BlockNumber,
+    /// Stage-specific checkpoint. None if stage uses only block-based checkpoints.
+    pub stage_checkpoint: Option<StageUnitCheckpoint>,
+}
+
+impl StageCheckpoint {
+    /// Creates a new [`StageCheckpoint`] with only `block_number` set.
+    pub fn new_with_block_number(block_number: BlockNumber) -> Self {
+        Self { block_number, ..Default::default() }
+    }
+}
+
+// TODO(alexey): ideally, we'd want to display block number + stage-specific metric (if available)
+//  in places like logs or traces
+impl Display for StageCheckpoint {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        Display::fmt(&self.block_number, f)
+    }
+}
+
+// TODO(alexey): add a merkle checkpoint. Currently it's hard because [`MerkleCheckpoint`]
+//  is not a Copy type.
+/// Stage-specific checkpoint metrics.
+#[derive_arbitrary(compact)]
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize, Deserialize)]
+pub enum StageUnitCheckpoint {
+    /// Saves the progress of transaction-indexed stages.
+    Transaction(TxNumber),
+    /// Saves the progress of AccountHashing stage.
+    Account(AccountHashingCheckpoint),
+    /// Saves the progress of StorageHashing stage.
+    Storage(StorageHashingCheckpoint),
+}
+
+impl Compact for StageUnitCheckpoint {
+    fn to_compact<B>(self, buf: &mut B) -> usize
+    where
+        B: BufMut + AsMut<[u8]>,
+    {
+        match self {
+            StageUnitCheckpoint::Transaction(data) => {
+                buf.put_u8(0);
+                1 + data.to_compact(buf)
+            }
+            StageUnitCheckpoint::Account(data) => {
+                buf.put_u8(1);
+                1 + data.to_compact(buf)
+            }
+            StageUnitCheckpoint::Storage(data) => {
+                buf.put_u8(2);
+                1 + data.to_compact(buf)
+            }
+        }
+    }
+
+    fn from_compact(buf: &[u8], _len: usize) -> (Self, &[u8])
+    where
+        Self: Sized,
+    {
+        match buf[0] {
+            0 => {
+                let (data, buf) = TxNumber::from_compact(&buf[1..], buf.len() - 1);
+                (Self::Transaction(data), buf)
+            }
+            1 => {
+                let (data, buf) = AccountHashingCheckpoint::from_compact(&buf[1..], buf.len() - 1);
+                (Self::Account(data), buf)
+            }
+            2 => {
+                let (data, buf) = StorageHashingCheckpoint::from_compact(&buf[1..], buf.len() - 1);
+                (Self::Storage(data), buf)
+            }
+            _ => unreachable!(),
```

Comment:
nit: add a description to unreachable stmt. useful to debug db incompatibility after breaking changes

Reply:
good call, also added to other places when unknown enum variant is encountered during the compact decoding: https://github.com/paradigmxyz/reth/pull/2701/commits/7ecbbdb511af7be747749be2c36c53e8becb3cb0